docs/DafnyRef/UserGuide.md

```diff
@@ -156,14 +156,227 @@ exit code.
+  <br><br>`assert forall i | 0 < i <= m :: P(i);` |  `assert forall i | 0 < i < m :: P(i);`<br>`assert forall i | i == m :: P(i);`<br>`assert forall i | 0 < i <= m :: P(i);`<br><br>
+  <br><br>`assert forall i | i == m :: P(m);` |  `assert P(m);`<br>`assert forall i | i == m :: P(i);`
+  `method m(i) returns (j: T)`<br>&nbsp;&nbsp;`  requires A(i)`<br>&nbsp;&nbsp;`  ensures B(i, j)`<br>`{`<br>&nbsp;&nbsp;`  ...`<br>`}`<br><br>`method n() {`<br>&nbsp;&nbsp;`  ...`<br><br><br>&nbsp;&nbsp;`  var x := m(a);`<br>&nbsp;&nbsp;`  assert P(x);` | `method m(i) returns (j: T)`<br>&nbsp;&nbsp;`  requires A(i)`<br>&nbsp;&nbsp;`  ensures B(i, j)`<br>`{`<br>&nbsp;&nbsp;`  ...`<br>`}`<br><br>`method n() {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A(k);`<br>&nbsp;&nbsp;`  assert forall x :: B(k, x) ==> P(x);`<br>&nbsp;&nbsp;`  var x := m(k);`<br>&nbsp;&nbsp;`  assert P(x);`
+
+### 24.8.2. Assertion batches {#sec-assertion-batches}
+
+To understand how to control verification,
+it is first useful to understand how Dafny verifies functions and methods.
+
+For every method (or function, constructor, etc.), Dafny extracts _assertions_, for example:
+
+* any explicit [`assert` statement](#sec-assert-statement) is _an assertion_[^precision-requires-clause].
+* A consecutive pair of lines in a [`calc` statement](#sec-calc-statement) forms _an assertion_.
+* Every function or method call with a [`requires` clause](#sec-requires-clause) yields _one assertion per requires clause_[^precision-requires-clause]
+  (special cases such as sequence indexing come with a special require clause that the index is within bounds).
+* Assignments `o.f := E;` yield an _assertion_ that `o.f` is allowed by the enclosing [`modifies` clause](#sec-loop-framing).
+* [Assign-such-that operators](#sec-update-and-call-statement) `x :| P(x)` yield an _assertion_ that `exists x :: P(x)`.
+* Every [`ensures` clause](#sec-ensures-clause) yields an _assertion_ at the end of the method and on every return.
+* Every [`decreases` clause](#sec-decreases-clause) yields an _assertion_ at either a call site or at the end of a while loop.
+* Every non-empty [subset type](#sec-subset-types) yields an _assertion_ that its witness satisfies the constraint.
+
+[^precision-requires-clause]: Dafny actually breaks things down further. For example, a precondition `requires A && B` or an assert statement `assert A && B;` turns into two assertions, more or less like `requires A requires B` and `assert A; assert B;`.
+
+It is useful to mentally visualize all these assertions as a list (an _assertion batch_) that roughly follows the order in the code[^complexity-path-encoding],
+except for `ensures` assertions that appear earlier in the code but, for verification purposes, would appear at the end.
+Thus, to prove or disprove a single assertion within this list, all the assumptions Dafny needs are 1) the global context,
+and 2) every preceding assumption (and previous assertions transformed in assumptions).
```

Comment:
This doesn't QUITE nail it for me, as I think it introduces "assertion batch" slightly too early. How about:

```suggestion
Each assertion is not just a boolean expression that must be true, but also depends on context, such as the definition of variables and other assertions that appear earlier in the control flow[^complexity-path-encoding]. The fundamental unit of verification in Dafny is an _assertion batch_: one or more assertions along with all the context they depend on. To reduce overhead, by default Dafny collects all the assertions in the body of a given method into a single batch that it sends to the verifier, which tries to prove it correct.
```

Reply:
I am doing a change that merges both what I did and what you suggest. Let me know if you like it.